### PR TITLE
PriceTicker : Part 228

### PR DESCRIPTION
**Pull request by cis-navjot-g**

### CODE REVIEW

File: src/api/commands/priceticker/PriceTickerRootCommand.ts [hunk 1]
```diff
@@ -9,24 +9,53 @@ import { BaseCommand } from '../BaseCommand';
...
     constructor(
-        @inject(Types.Core) @named(Core.Logger) public Logger: typeof LoggerType
+        @inject(Types.Core) @named(Core.Logger) public Logger: typeof LoggerType,
+        @inject(Types.Service) @named(Targets.Service.PriceTickerService) private priceTickerService: PriceTickerService
     ) {
         super(Commands.PRICETICKER_ROOT);
         this.log = new Logger(__filename);
     }
 
     @validate()
```
Comment (xludx) [inline src/api/commands/priceticker/PriceTickerRootCommand.ts]: add js doc
    /**
     * data.params[]:
     *  [0]: ...
     *
     * description
     *
     * @param {RpcRequest} data
     * @returns {Promise<ASD>}
     */

File: src/api/commands/priceticker/PriceTickerRootCommand.ts [hunk 2]
```diff
@@ -9,24 +9,53 @@ import { BaseCommand } from '../BaseCommand';
...
     public async execute( @request(RpcRequest) data: RpcRequest, rpcCommandFactory: RpcCommandFactory): Promise<any> {
-        return await this.executeNext(data, rpcCommandFactory);
+        if (data.params.length > 0) {
+            const currencies = data.params;
+            let returnData: any = [];
+            for (const currency of currencies) { // INR, USD
+                let searchData: any = await this.priceTickerService.search(currency);
```
Comment (xludx) [inline src/api/commands/priceticker/PriceTickerRootCommand.ts]: move the implementation to the service layer.
this.priceTickerService.getPriceTickers(currencies);

File: src/api/commands/priceticker/PriceTickerRootCommand.ts [hunk 3]
```diff
@@ -9,24 +9,53 @@ import { BaseCommand } from '../BaseCommand';
...
+            throw new MessageException('Currency can\'t be blank');
+        }
     }
 
     public help(): string {
-        return this.getName() + ' (fetch)';
+        return this.getName();
     }
 
     public description(): string {
```
Comment (xludx) [inline src/api/commands/priceticker/PriceTickerRootCommand.ts]: Command for fetching cryptocurrency price tickers.

File: src/api/commands/priceticker/PriceTickerRootCommand.ts [hunk 4]
```diff
@@ -9,24 +9,53 @@ import { BaseCommand } from '../BaseCommand';
...
+            return returnData;
+        } else {
+            throw new MessageException('Currency can\'t be blank');
+        }
     }
 
     public help(): string {
-        return this.getName() + ' (fetch)';
+        return this.getName();
     }
```
Comment (xludx) [inline src/api/commands/priceticker/PriceTickerRootCommand.ts]: '\<symbol\> [symbol...]'

File: src/api/requests/PriceTickerCreateRequest.ts [hunk 5]
```diff
@@ -13,16 +13,49 @@ export class PriceTickerCreateRequest extends RequestBody {
...
+    public crypto_last_updated: string;
+
     public crypto_price_currency: string;
 
+    public crypto_24h_volume_currency: string;
+
+    public crypto_market_cap_currency: string;
+
+
+
```
Comment (xludx) [inline src/api/requests/PriceTickerCreateRequest.ts]: adding ?convert=EUR to the request url should give you also prices and volumes in eur

File: test/black-box/PriceTickerRootCommand.test.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+describe('PriceTickerRootCommand', () => {
+
+    const testUtil = new BlackBoxTestUtil();
+    const method = Commands.PRICETICKER_ROOT.commandName;
+    beforeAll(async () => {
+        await testUtil.cleanDb();
+    });
+
+    test('Should fail because we want to create PriceTicker without passing currency', async () => {
+        const res = await rpc(method, []);
```
Comment (xludx) [inline test/black-box/PriceTickerRootCommand.test.ts]: we don't want to create a PriceTicker, we want to return/get it.

File: test/black-box/PriceTickerRootCommand.test.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    });
+
+    test('Should create records by passing two currency(one existing + one new)', async () => {
+        const res = await rpc(method, ['INR', 'EUR']);
+        res.expectJson();
+        res.expectStatusCode(200);
+        const result: any = res.getBody()['result'];
+        expect(result).toHaveLength(400);
+    });
+
```
Comment (xludx) [inline test/black-box/PriceTickerRootCommand.test.ts]: the purpose of this rpc api is not to create some data, but to return it. we are saving data to the db just for caching purposes.
"should return two pricetickers when passing two currencies"
although you can test if the caching mechanism works, but that's not the main purpose of the api that you should be testing.

Comment (xludx) [reply test/black-box/PriceTickerRootCommand.test.ts]: also INR, EUR will fail

File: src/api/services/PriceTickerService.ts [hunk 8]
```diff
@@ -92,14 +114,37 @@ export class PriceTickerService {
         await this.priceTickerRepo.destroy(id);
     }
```
Comment (xludx) [inline src/api/services/PriceTickerService.ts]: search method jsdoc should explain what the currency param should be exactly. should it be bitcoin or the symbol btc? Right now I cant tell that by looking at this code.

it should be symbol.
also, why is the return value a Collection when the search parameter implies that we are searching just for one currency?
search method should be renamed to getOneBySymbol()

File: src/api/services/PriceTickerService.ts [hunk 9]
```diff
@@ -136,12 +181,12 @@ export class PriceTickerService {
...
      */
-    public async needTobeUpdate(priceTicker: PriceTicker): Promise<boolean> {
+    private async needTobeUpdate(priceTicker: PriceTicker): Promise<boolean> {
         const diffMint = await this.checkDiffBtwDate(priceTicker['updatedAt']);
         return (diffMint > process.env.DATA_CHECK_DELAY) ? true : false;
     }
 
-    public async getAndCreateData(currency: string): Promise<any> {
+    private async getAndCreateData(currency: string): Promise<any> {
         const returnData: any = [];
```
Comment (xludx) [inline src/api/services/PriceTickerService.ts]: what is this returning?

File: src/api/services/PriceTickerService.ts [hunk 10]
```diff
@@ -92,14 +114,37 @@ export class PriceTickerService {
...
     /**
      * update PriceTicker
      *
      * @param priceTicker
      * @param apidata
      * @returns {Promise<any>}
      */
-    public async updatePriceTicker(priceTickerColl: any, currency: string): Promise<any> {
+    private async updatePriceTicker(priceTickerColl: any, currency: string): Promise<any> {
         // call api
```
Comment (xludx) [inline src/api/services/PriceTickerService.ts]: what is this returning?

File: test/integration/PriceTicker.test.ts [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+        expect(result.cryptoMarketCapCurrency).toBe(testDataUpdated.crypto_market_cap_currency);
+    });
+
+    test('Should delete the PriceTicker', async () => {
+        expect.assertions(1);
+        await priceTickerService.destroy(createdId);
+        await priceTickerService.findOne(createdId).catch(e =>
+            expect(e).toEqual(new NotFoundException(createdId))
+        );
+    });
```
Comment (xludx) [inline test/integration/PriceTicker.test.ts]: these tests are just mostly automatically generated for the default crud functionalities. 
you had a search function that you implemented and which is quite important for this implementation. yet there's no test for it here? how do you know that it works?

Comment (xludx) [reply test/integration/PriceTicker.test.ts]: I cannot see any tests making sure that the fetching from cmc works as expected and the prices are being updated on the db as expected

File: src/api/models/PriceTicker.ts [hunk 12]
```diff
@@ -43,13 +43,54 @@ export class PriceTicker extends Bookshelf.Model<PriceTicker> {
...
 
     public get CryptoPriceBtc(): string { return this.get('crypto_price_btc'); }
     public set CryptoPriceBtc(value: string) { this.set('crypto_price_btc', value); }
 
+    public get Crypto24HVolumeUsd(): string { return this.get('crypto_24_h_volume_usd'); }
+    public set Crypto24HVolumeUsd(value: string) { this.set('crypto_24_h_volume_usd', value); }
+
+    public get CryptoMarketCapUsd(): string { return this.get('crypto_market_cap_usd'); }
+    public set CryptoMarketCapUsd(value: string) { this.set('crypto_market_cap_usd', value); }
+
```
Comment (xludx) [inline src/api/models/PriceTicker.ts]: I cant see the eur prices here.

File: src/api/services/PriceTickerService.ts [hunk 13]
```diff
@@ -92,14 +114,37 @@ export class PriceTickerService {
         await this.priceTickerRepo.destroy(id);
     }
 
+    public async executePriceTicker(currencies: string[]): Promise<any> {
+        // const currencies = data.params;
+        let returnData: any = [];
+        for (const currency of currencies) { // INR, USD
+            let searchData: any = await this.search(currency);
```
[the remark this answers is not in the document]
Comment (cis-navjot-g) [reply src/api/services/PriceTickerService.ts]: done

File: src/api/services/PriceTickerService.ts [hunk 14]
```diff
@@ -92,14 +114,37 @@ export class PriceTickerService {
         await this.priceTickerRepo.destroy(id);
     }
 
+    public async executePriceTicker(currencies: string[]): Promise<any> {
+        // const currencies = data.params;
```
[the remark this answers is not in the document]
Comment (cis-navjot-g) [reply src/api/services/PriceTickerService.ts]: done

File: src/api/models/PriceTicker.ts (hunk 12 above)
Comment (cis-navjot-g) [reply src/api/models/PriceTicker.ts]: done

File: test/integration/PriceTicker.test.ts (hunk 11 above)
Comment (cis-navjot-g) [reply test/integration/PriceTicker.test.ts]: done

File: src/api/services/PriceTickerService.ts (hunk 8 above)
Comment (cis-navjot-g) [reply src/api/services/PriceTickerService.ts]: done

File: test/black-box/PriceTickerRootCommand.test.ts [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+        expect(res.error.error.message).toBe(`Currency can\'t be blank`);
+    });
+
+    test('Should create records by passing single currency', async () => {
+        const res = await rpc(method, ['INR']);
+        res.expectJson();
+        res.expectStatusCode(200);
+        const result: any = res.getBody()['result'];
+        expect(result).toHaveLength(200);
+    });
```
[the remark this answers is not in the document]
Comment (cis-navjot-g) [reply test/black-box/PriceTickerRootCommand.test.ts]: done

File: src/api/services/PriceTickerService.ts (hunk 9 above)
Comment (cis-navjot-g) [reply src/api/services/PriceTickerService.ts]: I have removed that and put single function for update and check

File: src/api/commands/priceticker/PriceTickerRootCommand.ts [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+     *
+     * @param data
+     * @returns {Promise<Bookshelf.Collection<PriceTicker>>}
+     */
+    @validate()
+    public async execute( @request(RpcRequest) data: RpcRequest, rpcCommandFactory: RpcCommandFactory): Promise<Bookshelf.Collection<PriceTicker>> {
+        if (data.params.length > 0) {
+            let returnData: any = [];
+            returnData = await this.priceTickerService.executePriceTicker(data.params);
+            return returnData;
```
Comment (xludx) [inline src/api/commands/priceticker/PriceTickerRootCommand.ts]: rename executePriceTicker to getPriceTickers. current name implies that we somehow execute a single PriceTicker which is not the case.

File: src/api/models/PriceTicker.ts [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+export class PriceTicker extends Bookshelf.Model<PriceTicker> {
+
+    public static async fetchById(value: number, withRelated: boolean = true): Promise<PriceTicker> {
+        if (withRelated) {
+            return await PriceTicker.where<PriceTicker>({ id: value }).fetch({
+                withRelated: [
+                    // TODO:
+                    // 'PriceTickerRelated',
+                    // 'PriceTickerRelated.Related'
+                ]
```
Comment (xludx) [inline src/api/models/PriceTicker.ts]: these comments are not needed.

File: src/api/services/PriceTickerService.ts [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+        // TODO: extract and remove related models from request
+        // const priceTickerRelated = body.related;
+        // delete body.related;
+
+        // If the request body was valid we will create the priceTicker
+        const priceTicker = await this.priceTickerRepo.create(body);
+
+        // TODO: create related models
+        // priceTickerRelated._id = priceTicker.Id;
+        // await this.priceTickerRelatedService.create(priceTickerRelated);
```
Comment (xludx) [inline src/api/services/PriceTickerService.ts]: no need for these comments

File: test/.env.test.example [hunk 19]
```diff
@@ -65,3 +65,8 @@ DB_SEEDS_DIR="./src/database/seeds"
 # Auth0
 #
 AUTH0_HOST="http://localhost:3333"
+
+#
+# PriceTicker
+#
+DATA_CHECK_DELAY=0.05
```
Comment (xludx) [inline test/.env.test.example]: change this to seconds

File: src/api/commands/priceticker/PriceTickerRootCommand.ts (hunk 16 above)
Comment (cis-navjot-g) [reply src/api/commands/priceticker/PriceTickerRootCommand.ts]: done

File: src/api/models/PriceTicker.ts (hunk 17 above)
Comment (cis-navjot-g) [reply src/api/models/PriceTicker.ts]: removed

File: src/api/services/PriceTickerService.ts (hunk 18 above)
Comment (cis-navjot-g) [reply src/api/services/PriceTickerService.ts]: removed

File: test/.env.test.example (hunk 19 above)
Comment (cis-navjot-g) [reply test/.env.test.example]: private async checkDiffBtwDate(timestamp: Date): Promise<number> {
        const current: any = new Date();
        const ticker: any = new Date(timestamp);
        return (current - ticker) / 1000;
    }

now this functions return in seconds

**Pull request merged**